<a href="https://colab.research.google.com/github/kapilpriya01/priya-kapil/blob/master/Rf_DT/RF_DT_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd


In [56]:
data = pd.DataFrame()
data = pd.read_csv("https://raw.githubusercontent.com/insaid2018/CDF-Certification-Quiz-Data/master/Bank1/02-Data.csv")
data.head()

ID  Age  Experience  ...  CD Account  Online  CreditCard
0   1   25           1  ...           0       0           0
1   2   45          19  ...           0       0           0
2   3   39          15  ...           0       0           0
3   4   35           9  ...           0       0           0
4   5   35           8  ...           0       0           1

[5 rows x 13 columns]

In [57]:
data.shape

(5000, 13)

In [58]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   Family              5000 non-null   int64  
 5   CCAvg               5000 non-null   float64
 6   Education           5000 non-null   int64  
 7   Mortgage            5000 non-null   int64  
 8   Personal Loan       5000 non-null   int64  
 9   Securities Account  5000 non-null   int64  
 10  CD Account          5000 non-null   int64  
 11  Online              5000 non-null   int64  
 12  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(12)
memory usage: 507.9 KB


In [59]:
X = data.loc[: , data.columns != "CreditCard"]
X.head()

ID  Age  Experience  ...  Securities Account  CD Account  Online
0   1   25           1  ...                   1           0       0
1   2   45          19  ...                   1           0       0
2   3   39          15  ...                   0           0       0
3   4   35           9  ...                   0           0       0
4   5   35           8  ...                   0           0       0

[5 rows x 12 columns]

In [0]:
y = data.CreditCard

In [0]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.30, random_state = 2018)

In [62]:
print(X_train.shape)
print(y_train.shape)

(3500, 12)
(3500,)


In [63]:
print(X_test.shape)
print(y_test.shape)

(1500, 12)
(1500,)


In [0]:
from sklearn import tree

In [28]:
model = tree.DecisionTreeRegressor(random_state=2019)
model.fit(X_train,y_train)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=2019, splitter='best')

In [0]:
y_pred_train = model.predict(X_train)
y_pred_test = model.predict (X_test)

In [68]:
from sklearn.metrics import accuracy_score
print("Accuracy score",accuracy_score(y_test,y_pred_test))

Accuracy score 0.6326666666666667


In [0]:
from sklearn.metrics import confusion_matrix

In [70]:
confusion_matrix_dt = confusion_matrix(y_test,y_pred_test)
confusion_matrix_dt

array([[782, 296],
       [255, 167]])

Thus

* TP: 167
* TN: 782
* FP = 296
* FN = 255



In [0]:
from sklearn.ensemble import RandomForestClassifier

In [91]:
from sklearn.model_selection import train_test_split

X_train1, X_test1, y_train1, y_test1 = train_test_split(X,y,test_size = 0.30, random_state = 2018)

rf_model = RandomForestClassifier(random_state=0)
rf_model.fit(X_train1,y_train1)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [0]:
y_pred_train1 = rf_model.predict(X_train1)
y_pred_test1 = rf_model.predict (X_test1)

In [0]:
from sklearn.metrics import accuracy_score

In [94]:
print("Accuracy score from random forests",accuracy_score(y_test1,y_pred_test1))

Accuracy score from random forests 0.7466666666666667


In [95]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test1,y_pred_test1)
confusion_matrix

array([[1034,   44],
       [ 336,   86]])

Thus 
* TP: 86
* TN: 1034
* FP = 44
* FN = 336

Thus DTs is better here as here FNs are lower